app/src/main/java/com/example/itbook/ui/search/SearchFragment.kt

```diff
@@ -0,0 +1,135 @@
+import com.example.itbook.data.source.local.dao.BookDaoImp
+import com.example.itbook.data.source.remote.BookRemoteHandler
+import com.example.itbook.data.source.remote.BooksRemoteDataSource
+import com.example.itbook.ui.adapter.PreviewBookAdapter
+import com.example.itbook.ui.detailbook.DetailBookFragment
+import com.example.itbook.ui.dialog.LoadingDialogFragment
+import com.example.itbook.utils.closeKeyboard
+import com.example.itbook.utils.openKeyboard
+import com.example.itbook.utils.showError
+import kotlinx.android.synthetic.main.fragment_search.*
+import org.json.JSONException
+
+class SearchFragment : BaseFragment(), SearchContract.View {
+    override val layoutResource: Int = R.layout.fragment_search
+
+    private var presenter: SearchPresenter? = null
+    private var booksRepository: BooksRepository? = null
+    private val previewBookAdapter: PreviewBookAdapter = PreviewBookAdapter(this::onBookClick)
+    private var loadingDialogFragment: LoadingDialogFragment? = null
+    private var books = mutableListOf<Book>()
+
+    override fun initViews() {
+        editSearch.requestFocus()
+        activity?.let { editSearch.openKeyboard(it) }
+        recyclerSearchBooks.adapter = previewBookAdapter
+        loadingDialogFragment = LoadingDialogFragment()
+        loadingDialogFragment?.isCancelable = false
+        isSetup = false
+    }
+
+    override fun initData() {
+        activity?.let {
+            booksRepository = BooksRepository.getInstance(
+                BooksRemoteDataSource.getInstance(BookRemoteHandler()),
+                BooksLocalDataSource.getInstance(BookDaoImp(BooksDatabase.getInstance(it)))
+            )
+        }
+    }
+
+    override fun initListeners() {
+        editSearch.setOnEditorActionListener(onEditorActionListener(editSearch))
+        imageSearch.setOnClickListener { onSearchListener(editSearch) }
+        imageCancelSearch.setOnClickListener { onCancelClickListener() }
+    }
+
+    override fun showSearchResult(books: List<Book>) {
+        if (books.isNotEmpty()) {
+            val booksSize = this.books.size - 1
+            val insertCount = books.size - this.books.size
+
+            previewBookAdapter.apply {
+                updateData(books.toMutableList())
+                notifyItemRangeInserted(booksSize, insertCount)
+            }
+            this.books = books.toMutableList()
+        } else {
+            previewBookAdapter.apply {
+                updateData(mutableListOf())
+                notifyDataSetChanged()
+            }
+        }
+    }
+
+    override fun resetResult() {
+        previewBookAdapter.apply {
+            updateData(mutableListOf())
+            notifyDataSetChanged()
+        }
+    }
+
+    override fun showError(error: Exception?) {
+        var errorMessage = ""
+        when (error) {
+            is JSONException -> {
+                errorMessage = resources.getString(R.string.error_internet_not_connection)
+            }
```

Comment:
sao lỗi JSON lại trả về message của internet connection?

Reply:
dạ tại nếu lỗi mạng thì không load được -> không convert được sang json-> lỗi. nên em để vậy ạ.

Reply:
2 lỗi này khác hoàn toàn nhau nên phải xử lý khác nhau chứ